console/src/state/resources.rs

```diff
@@ -0,0 +1,322 @@
+            }),
+            Self::Total => resources
+                .sort_unstable_by_key(|resource| resource.upgrade().map(|r| r.borrow().total(now))),
+        }
+    }
+}
+
+impl TryFrom<usize> for SortBy {
+    type Error = ();
+    fn try_from(idx: usize) -> Result<Self, Self::Error> {
+        match idx {
+            idx if idx == Self::Rid as usize => Ok(Self::Rid),
+            idx if idx == Self::Kind as usize => Ok(Self::Kind),
+            idx if idx == Self::ConcreteType as usize => Ok(Self::ConcreteType),
+            idx if idx == Self::Target as usize => Ok(Self::Target),
+            idx if idx == Self::Total as usize => Ok(Self::Total),
+            _ => Err(()),
+        }
+    }
+}
+
+impl view::SortBy for SortBy {
+    fn as_usize(&self) -> usize {
+        *self as usize
+    }
+}
+
+impl ResourcesState {
+    pub(crate) fn take_new_resources(&mut self) -> impl Iterator<Item = ResourceRef> + '_ {
+        self.new_resources.drain(..)
+    }
+
+    pub(crate) fn update_resources(
+        &mut self,
+        styles: &view::Styles,
+        update: proto::resources::ResourceUpdate,
+        metas: &HashMap<u64, Metadata>,
+        clear_new_list: bool,
+    ) {
+        let mut stats_update = update.stats_update;
+        let new_list = &mut self.new_resources;
+        if clear_new_list {
+            new_list.clear();
+        }
+
+        let new_resources = update.new_resources.into_iter().filter_map(|mut resource| {
+            if resource.id.is_none() {
+                tracing::warn!(?resource, "skipping resource with no id");
+            }
+
+            let meta_id = match resource.metadata.as_ref() {
+                Some(id) => id.id,
+                None => {
+                    tracing::warn!(?resource, "resource has no metadata ID, skipping");
+                    return None;
+                }
+            };
+            let meta = match metas.get(&meta_id) {
+                Some(meta) => meta,
+                None => {
+                    tracing::warn!(?resource, meta_id, "no metadata for resource, skipping");
+                    return None;
+                }
+            };
+
+            let id = resource.id?.id;
+            let stats = ResourceStats::from_proto(stats_update.remove(&id)?, meta, styles);
+            let kind = resource.kind?.into();
+
+            // remove cargo part of the file path
+            resource.location = resource.location.map(|mut l| {
+                if let Some(file) = l.file.take() {
+                    let truncated = truncate_registry_path(file);
+                    l.file = Some(truncated);
+                }
+                l
+            });
+            let resource = Resource {
+                id,
+                kind,
+                stats,
+                target: meta.target.clone(),
+                concrete_type: resource.concrete_type,
+                meta_id,
+                location: resource.location,
```

Comment:
nit: it's a little bit strange to me that we mutate the protobuf `resource`'s location field in place, and then add it to the new `Resource` struct, instead of just doing this:
```suggestion
            let location = resource.location.take().map(|mut l| {
                if let Some(file) = l.file.take() {
                    let truncated = truncate_registry_path(file);
                    l.file = Some(truncated);
                }
                l
            });
            let resource = Resource {
                id,
                kind,
                stats,
                target: meta.target.clone(),
                concrete_type: resource.concrete_type,
                meta_id,
                location,
```
but, not a major issue!

Reply:
Yes... that was a bit strange for sure